frame/contracts/src/lib.rs

```diff
@@ -320,15 +320,30 @@ pub mod pallet {
 		T::AccountId: UncheckedFrom<T::Hash>,
 		T::AccountId: AsRef<[u8]>,
 	{
+		fn on_idle(_block: T::BlockNumber, remaining_weight: Weight) -> Weight {
+			Storage::<T>::process_deletion_queue_batch(remaining_weight)
+				.saturating_add(T::WeightInfo::on_idle())
+		}
+
 		fn on_initialize(_block: T::BlockNumber) -> Weight {
-			// We do not want to go above the block limit and rather avoid lazy deletion
-			// in that case. This should only happen on runtime upgrades.
-			let weight_limit = T::BlockWeights::get()
-				.max_block
-				.saturating_sub(System::<T>::block_weight().total())
-				.min(T::DeletionWeightLimit::get());
-			Storage::<T>::process_deletion_queue_batch(weight_limit)
-				.saturating_add(T::WeightInfo::on_initialize())
+			// We want to process the deletion_queue in the on_idle hook. Only in the case
+			// that the queue length has reached its maximal depth, we process it here
+			let queue_depth = T::DeletionQueueDepth::get() as usize;
+			let queue_len = <DeletionQueue<T>>::decode_len().unwrap_or(0);
+			let queue_full = queue_len >= queue_depth;
+			if queue_full {
```

Comment:
I think we can save this binding. Should be easy enough to understand.

```suggestion
			let max_len = T::DeletionQueueDepth::get() as usize;
			let queue_len = <DeletionQueue<T>>::decode_len().unwrap_or(0);
			if queue_len >= max_len {
```

Reply:
I agree, its explicit enough.